drivers/timer/apic_tsc.c

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2021 Intel Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <drivers/timer/system_timer.h>
+#include <sys_clock.h>
+#include <spinlock.h>
+#include <drivers/interrupt_controller/loapic.h>
+
+#define IA32_TSC_DEADLINE_MSR 0x6e0
+#define IA32_TSC_ADJUST_MSR   0x03b
+
+#define CYC_PER_TICK (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC \
+		      / (uint64_t) CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+struct apic_timer_lvt {
+	uint8_t vector   : 8;
+	uint8_t unused0  : 8;
+	uint8_t masked   : 1;
+	enum { ONE_SHOT, PERIODIC, TSC_DEADLINE } mode : 2;
+	uint32_t unused2 : 13;
+};
+
+static struct k_spinlock lock;
+static uint64_t last_announce;
+static union { uint32_t val; struct apic_timer_lvt lvt; } lvt_reg;
+
+static ALWAYS_INLINE uint64_t rdtsc(void)
+{
+	uint32_t hi, lo;
+
+	__asm__ volatile("rdtsc" : "=d"(hi), "=a"(lo));
+	return lo + (((uint64_t)hi) << 32);
+}
+
+static void isr(const void *arg)
```

Comment:
This is one showcase here: note that this interrupt handler contains basically no code.  Compute elapsed ticks with a subtract and divide and announce them.